File: src/data/courses/dsa/two-pointers-sliding-window/sliding-window-variable-longest/longest-subarray-sum-k/code/cpp/longest_subarray_sum_k.cpp
// ============================================================
// Longest Subarray with Sum K
// ============================================================

#include <vector>
#include <unordered_map>
using namespace std;

// ============================================================
// Solution 1: Brute Force
// Time: O(N²) | Space: O(1)
// ============================================================
class Solution1 {
public:
    int longestSubarray(vector<int>& nums, int k) {
        int maxLen = 0;
        for (int i = 0; i < nums.size(); i++) {
            int sum = 0;
            for (int j = i; j < nums.size(); j++) {
                sum += nums[j];
                if (sum == k) maxLen = max(maxLen, j - i + 1);
            }
        }
        return maxLen;
    }
};

// ============================================================
// Solution 2: Prefix Sum + Hash Map (Optimal)
// Time: O(N) | Space: O(N)
// ============================================================
class Solution2 {
public:
    int longestSubarray(vector<int>& nums, int k) {
        unordered_map<long, int> firstOccurrence;
        firstOccurrence[0] = -1; // prefix sum 0 at index -1

        long prefixSum = 0;
        int maxLen = 0;

        for (int j = 0; j < nums.size(); j++) {
            prefixSum += nums[j];

            // Check if (prefixSum - k) was seen before
            if (firstOccurrence.count(prefixSum - k)) {
                maxLen = max(maxLen, j - firstOccurrence[prefixSum - k]);
            }

            // Store FIRST occurrence only (for maximum length)
            if (!firstOccurrence.count(prefixSum)) {
                firstOccurrence[prefixSum] = j;
            }
        }

        return maxLen;
    }
};

File: src/data/courses/dsa/two-pointers-sliding-window/sliding-window-variable-longest/longest-subarray-sum-k/code/python/longest_subarray_sum_k.py
# ============================================================
# Longest Subarray with Sum K
# ============================================================

from typing import List

# ============================================================
# Solution 1: Brute Force
# Time: O(N²) | Space: O(1)
# ============================================================
class Solution1:
    def longestSubarray(self, nums: List[int], k: int) -> int:
        max_len = 0
        for i in range(len(nums)):
            total = 0
            for j in range(i, len(nums)):
                total += nums[j]
                if total == k:
                    max_len = max(max_len, j - i + 1)
        return max_len

# ============================================================
# Solution 2: Prefix Sum + Hash Map (Optimal)
# Time: O(N) | Space: O(N)
# ============================================================
class Solution2:
    def longestSubarray(self, nums: List[int], k: int) -> int:
        first_occurrence = {0: -1}  # prefix_sum → first index
        prefix_sum = 0
        max_len = 0

        for j in range(len(nums)):
            prefix_sum += nums[j]

            # Check if complement prefix sum was seen
            if prefix_sum - k in first_occurrence:
                max_len = max(max_len, j - first_occurrence[prefix_sum - k])

            # Store first occurrence only (for max length)
            if prefix_sum not in first_occurrence:
                first_occurrence[prefix_sum] = j

        return max_len
